LiS/man/man4/nullmod.4.man
'\" rtp
.\" -*- nroff -*- vim: ft=nroff nocin nosi
.\"
.\" @(#) $Id: nullmod.4.man,v 0.9.2.3 2008-04-29 08:33:19 brian Exp $
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.  The author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.\" Last Modified $Date: 2008-04-29 08:33:19 $ by $Author: brian $
.\"
.\" -------------------------------------------------------------------------
.\"
.\" $Log: nullmod.4.man,v $
.\" Revision 0.9.2.3  2008-04-29 08:33:19  brian
.\" - update headers for Affero release
.\"
.\" Revision 0.9.2.2  2006/04/05 09:42:35  brian
.\" - moved .so line and updated headers
.\"
.\" =========================================================================
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database lis.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
.R2
.so lis.macros
.\"
.\"
.TH NULLMOD 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Modules"
.\"
.\"
.SH NAME
.B nullmod
\- \fISTREAMS\fR null module
.\"
.\"
.SH SYNOPSIS
.PP
.HP 8
.BI "int " fd ;
.HP 8
.BI "int open(" \*(lq\fR/dev/nuls\fP\*(rq ", " \fR0\fP );
.HP 8
.BI "int ioctl(" fd ", " I_PUSH ", " \*(lqnullmod\*(rq );
.\"
.\"
.SH DESCRIPTION
.PP
.B nullmod
is a
.I STREAMS
module that is pushed over any
.IR Stream ,
.IR STREAMS -based
.BR pipe (4)
or
.IR STREAMS -based
.BR fifo (4).
.PP
Aside from being a
.I STREAMS
module,
.B nullmod
performs no actions.  All messages passed from the
.I Stream head
toward the
.B nullmod
module are passed on downstream from the
.B nullmod
write-side
.B put
procedure (with
.BR putnext (9));
all messages received from the driver,
are passed on upstream from the
read-side
.B put
procedure.  The
.B nullmod
module has no write- or read-side
.B service
procedure.
.\"
.\"
.SH USAGE
.PP
The primary purpose of
.B nullmod
is as a test module for use with the
.BR test-streams (8)
conformance test program.  The other purpose of
.B nullmod
is as an example skeleton for a
.I STREAMS
pushable module.
.\"
.\"
.SH NOTICES
.PP
.BR nullmod
is implementation-specific to the
.B @PACKAGE_TITLE@
package.
.\"
.\"
.SH EXAMPLES
.PP
The source code for
.B nullmod
serves as an example skeleton for a
.I STREAMS
pushable module.
.\"
.\"
.SH MODULES
.PP
.B nullmod
is a
.I STREAMS
module named
.IR \*(lqnullmod\*(rq .
.\"
.\"
.SH "SEE ALSO"
.PP
.BR open (2s),
.BR ioctl (2),
.BR streamio (7),
.BR putnext (9).
.\"
.\"
.SH BUGS
.PP
.B nullmod
has no known bugs.
.\"
.\"
.SH COMPATIBILITY
.PP
.B nullmod
is compatible with the null module described in the
.IR "SVR 4.0" ", \*(lq" "STREAMS Programmer's Guide" ",\*(rq"
.[
svr4spg
%P Section 6.0.9.1, "Null Module Example"
.]
with the following portability considerations:
.IP \(em \w'\(em\(em'u
.B nullmod
is specific to
.BR "@PACKAGE_TITLE@" ,
and will not be used by portable
.I STREAMS
modules, drivers and applications programs.
.IP \(em
The
.I SVR 4.0 SPG
.[
svr4spg
.]
does not name this module.  The name
.B nullmod
was chosen for
.BR "@PACKAGE_TITLE@" .
.IP \(em
.B nullmod
is used as an example skeleton module and for use by the
.BR test-streams (8)
conformance test program.  It will not be used by portable
.I STREAMS
modules, drivers or applications.
.PP
See
.BR STREAMS (9)
for additional compatibility information.
.\"
.\"
.SH CONFORMANCE
.PP
.IR "SVR 4.0" .
.[
svr4spg
.]
Conformance is verified using the
.BR test-nullmod (8)
verification test suite.
.\"
.\"
.SH HISTORY
.PP
.B nullmod
first appeared in
.IR "SVR 4.0" .
.\"
.\"
.[
$LIST$
.]
.TI


NEWS
OpenSS7 Master Package -- history of user visible changes.  2008-05-03
$Id: NEWS,v 0.9.2.53 2008-05-03 21:22:25 brian Exp $
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2000  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
OpenSS7 Master Package package.

 Major changes for release openss7-0.9.2.G
------------------------------------------

This is the sixth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

The `OpenSS7 Master Package' is not released as often as the
sub-packages.  As sub-packages are released more often, to rebuild the
master package with a new sub-package release, simply replace the
directory to which the sub-package belongs with the unpacked
sub-package release and then rebuild the master package.  This release
provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Documented `lib32gcc1' problem on Ubuntu.

   - Noted problem running under XEN kernels.  XEN kernels are not yet
     supported.

   - License upgrade to AGPL v3.

   - Cross-compile fixes (strap out AC_FUNC_REALLOC macro when
     cross-compiling).

   - Ability to strap out major documentation build and installation
     primarily for embedded targets.

   - Improvements to common build process for embedded and
     cross-compile targets.

   - Support for flex 2.5.33  in maintainer mode.

   - Added the following sub-packages to the master build (and relase):

         - `strx25-0.9.2.1'


_This is a public stable production grade release of the package: it
deprecates previous releases.  Please upgrade to the current release
before reporting bugs._

As with other OpenSS7 releases, this release configures, compiles,
installs and builds RPMs and DEBs for a wide range of Linux 2.4 and 2.6
RPM- and DPKG-based distributions, and can be used on production
kernels without patching or recompiling the kernel.

This package is publicly released under the `GNU Affero General Public
License Version 3'.  The release is available as an `autoconf' tarball,
SRPM, DSC, and set of binary RPMs and DEBs.  See the downloads page
(http://www.openss7.org/download.html) for the `autoconf' tarballs,
SRPMs and DSCs.  For tarballs, SRPMs, DSCs and binary RPMs and DEBs,
see the openss7 package page (http://www.openss7.org/openss7_pkg.html).

See `http://www.openss7.org/codefiles/openss7-0.9.2.G/ChangeLog' and
`http://www.openss7.org/codefiles/openss7-0.9.2.G/NEWS' in the release
for more information.  Also, see the `openss7.pdf' manual in the
release (also in html `http://www.openss7.org/openss7_manual.html').

 Major changes for release openss7-0.9.2.F
------------------------------------------

This is the fifth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

The `OpenSS7 Master Package' is not released as often as the
sub-packages.  As sub-packages are released more often, to rebuild the
master package with a new sub-package release, simply replace the
directory to which the sub-package belongs with the unpacked
sub-package release and then rebuild the master package.  This release
provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Support build on openSUSE 10.2.

   - Support build on Fedora 7 and 2.6.21 kernel.

   - Support build on CentOS 5.0 (RHEL5).

   - Support build on Ubuntu 7.04.

   - Updated to gettext 0.16.1.

   - Changes to support build on 2.6.20-1.2307.fc5 and
     2.6.20-1.2933.fc6 kernel.

   - Supports build on Fedora Core 6.

   - Support for recent distributions and tool chains.


 Major changes for release openss7-0.9.2.E
------------------------------------------

This is the fourth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

It is unlikely that the `OpenSS7 Master Package' will be released as
frequently as before.  Sub-packages will be released more often.  To
rebuild the master package with a new sub-package release, simply
replace the directory to which the sub-package belongs with the
unpacked sub-package release and then rebuild the master package.  This
release provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Addition of the `osr61' sub-package that contains Dialogic(R) Open
     System Release 6.1 version 239 GPL drivers.

   - A few minor corrections to the common build process.

   - Support for autoconf 2.61, automake 1.10 and gettext 0.16.

   - Support for Ubuntu 6.10 distribution and bug fixes for i386
     kernels.

   - The package now looks for subpackages with a version number as
     unpacked by separate tarball.


 Major changes for release openss7-0.9.2.D
------------------------------------------

This is the fourth public release of the OpenSS7 Master Package.  The
sub-packages have been reorganized for this release.  See `README' in
the release for a sub-package listing.  Aside from sub-package
reorganization, the major difference from previous release is that this
release no longer contains `LiS'.  Too many of the sub-packages will
not even build against `LiS' because of its Stream head deficiencies.


LiS/drivers/str/nullmod.c
/*****************************************************************************

 @(#) $RCSfile: nullmod.c,v $ $Name:  $($Revision: 0.9.2.14 $) $Date: 2008-04-29 08:33:11 $

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified $Date: 2008-04-29 08:33:11 $ by $Author: brian $

 -----------------------------------------------------------------------------

 $Log: nullmod.c,v $
 Revision 0.9.2.14  2008-04-29 08:33:11  brian
 - update headers for Affero release

 Revision 0.9.2.13  2007/08/15 04:57:58  brian
 - GPLv3 updates

 Revision 0.9.2.12  2007/08/14 10:46:56  brian
 - GPLv3 header update

 *****************************************************************************/

#ident "@(#) $RCSfile: nullmod.c,v $ $Name:  $($Revision: 0.9.2.14 $) $Date: 2008-04-29 08:33:11 $"

static char const ident[] =
    "$RCSfile: nullmod.c,v $ $Name:  $($Revision: 0.9.2.14 $) $Date: 2008-04-29 08:33:11 $";

/*
 *  This is NULLMOD a STREAMS null module that performs no actions other than acting as a STREAMS
 *  module.  Its purpose is primarily for testing and for serviing as an example of the skeleton of
 *  a STREAMS module.
 *
 *  This is an absurdly simple module.
 *
 *  In addition, this module provide a specialized set of input-output controls for testing and
 *  verifying various internal STREAMS functions.
 */

#include <linux/autoconf.h>
#include <linux/version.h>
#include <linux/module.h>
#include <linux/init.h>

#include <sys/kmem.h>
#include <sys/stream.h>
#include <sys/strconf.h>
#ifdef LFS
#include <sys/strsubr.h>
#endif
#include <sys/ddi.h>

#ifdef LFS
#include "sys/config.h"
#endif

#define NULLMOD_DESCRIP		"UNIX SYSTEM V RELEASE 4.2 FAST STREAMS FOR LINUX"
#define NULLMOD_COPYRIGHT	"Copyright (c) 1997-2008 OpenSS7 Corporation.  All Rights Reserved."
#define NULLMOD_REVISION	"LfS $RCSfile: nullmod.c,v $ $Name:  $($Revision: 0.9.2.14 $) $Date: 2008-04-29 08:33:11 $"
#define NULLMOD_DEVICE		"SVR 4.2 Null Module (NULLMOD) for STREAMS"
#define NULLMOD_CONTACT		"Brian Bidulock <bidulock@openss7.org>"
#define NULLMOD_LICENSE		"GPL"
#define NULLMOD_BANNER		NULLMOD_DESCRIP		"\n" \
				NULLMOD_COPYRIGHT	"\n" \
				NULLMOD_REVISION	"\n" \
				NULLMOD_DEVICE		"\n" \
				NULLMOD_CONTACT		"\n"
#define NULLMOD_SPLASH		NULLMOD_DEVICE		" - " \
				NULLMOD_REVISION	"\n"

#if defined LIS && defined MODULE
#define CONFIG_STREAMS_NULLMOD_MODULE MODULE
#endif

#ifdef CONFIG_STREAMS_NULLMOD_MODULE
MODULE_AUTHOR(NULLMOD_CONTACT);
MODULE_DESCRIPTION(NULLMOD_DESCRIP);
MODULE_SUPPORTED_DEVICE(NULLMOD_DEVICE);
MODULE_LICENSE(NULLMOD_LICENSE);
#if defined MODULE_ALIAS
MODULE_ALIAS("streams-nullmod");
#endif
#endif

#ifndef CONFIG_STREAMS_NULLMOD_NAME
#ifdef LIS
#define CONFIG_STREAMS_NULLMOD_NAME NULLMOD__MOD_NAME
#else
#error "CONFIG_STREAMS_NULLMOD_NAME must be defined."
#endif
#endif
#ifndef CONFIG_STREAMS_NULLMOD_MODID
#ifdef LIS
#define CONFIG_STREAMS_NULLMOD_MODID NULLMOD__ID
#else
#error "CONFIG_STREAMS_NULLMOD_MODID must be defined."
#endif
#endif

modID_t modid = CONFIG_STREAMS_NULLMOD_MODID;

#ifndef module_param
MODULE_PARM(modid, "h");
#else
module_param(modid, ushort, 0444);
#endif
MODULE_PARM_DESC(modid, "Module ID for NULLMOD.");

#ifdef MODULE_ALIAS
#ifdef LFS
MODULE_ALIAS("streams-modid-" __stringify(CONFIG_STREAMS_NULLMOD_MODID));
MODULE_ALIAS("streams-module-nullmod");
#endif
#endif

#ifdef LIS
#define STRMINPSZ	0
#define STRMAXPSZ	4096
#define STRHIGH		5120
#define STRLOW		1024
#endif

STATIC struct module_info nullmod_minfo = {
	.mi_idnum = CONFIG_STREAMS_NULLMOD_MODID,
	.mi_idname = CONFIG_STREAMS_NULLMOD_NAME,
	.mi_minpsz = STRMINPSZ,
	.mi_maxpsz = STRMAXPSZ,
	.mi_hiwat = STRHIGH,
	.mi_lowat = STRLOW,
};

STATIC struct module_stat nullmod_rstat __attribute__ ((__aligned__(SMP_CACHE_BYTES)));
STATIC struct module_stat nullmod_wstat __attribute__ ((__aligned__(SMP_CACHE_BYTES)));

/* 
 *  -------------------------------------------------------------------------
 *
 *  Test and Verification Cases
 *
 *  -------------------------------------------------------------------------
 */

#ifdef LIS
union ioctypes {
	struct iocblk iocblk;
	struct copyreq copyreq;
	struct copyresp copyresp;
};
#endif

/*
 *  Test Case 1: allocb()  Test allocation and freeing of message blocks.
 */
STATIC int
testcase_1(union ioctypes *ioc, mblk_t *dp)
{
	ioc->iocblk.ioc_rval = 0;
	ioc->iocblk.ioc_error = 0;
	ioc->iocblk.ioc_count = 0;
	return (0);
}

STATIC int
testcase_2(union ioctypes *ioc, mblk_t *dp)
{
	ioc->iocblk.ioc_rval = 0;
	ioc->iocblk.ioc_error = 0;
	ioc->iocblk.ioc_count = 0;
	return (0);
}

STATIC int
testcase_3(union ioctypes *ioc, mblk_t *dp)
{
	ioc->iocblk.ioc_rval = 0;
	ioc->iocblk.ioc_error = 0;
	ioc->iocblk.ioc_count = 0;
	return (0);
}

STATIC int
testcase_4(union ioctypes *ioc, mblk_t *dp)
{
	ioc->iocblk.ioc_rval = 0;
	ioc->iocblk.ioc_error = 0;
	ioc->iocblk.ioc_count = 0;
	return (0);
}

STATIC int
testcase_5(union ioctypes *ioc, mblk_t *dp)
{
	ioc->iocblk.ioc_rval = 0;
	ioc->iocblk.ioc_error = 0;
	ioc->iocblk.ioc_count = 0;
	return (0);
}

STATIC int
testcase_6(union ioctypes *ioc, mblk_t *dp)
{
	ioc->iocblk.ioc_rval = 0;
	ioc->iocblk.ioc_error = 0;
	ioc->iocblk.ioc_count = 0;
	return (0);
}

STATIC int
testcase_7(union ioctypes *ioc, mblk_t *dp)
{
	ioc->iocblk.ioc_rval = 0;
	ioc->iocblk.ioc_error = 0;
	ioc->iocblk.ioc_count = 0;
	return (0);
}

STATIC int
testcase_8(union ioctypes *ioc, mblk_t *dp)
{
	ioc->iocblk.ioc_rval = 0;
	ioc->iocblk.ioc_error = 0;
	ioc->iocblk.ioc_count = 0;
	return (0);
}

STATIC int
testcase_9(union ioctypes *ioc, mblk_t *dp)
{
	ioc->iocblk.ioc_rval = 0;
	ioc->iocblk.ioc_error = 0;
	ioc->iocblk.ioc_count = 0;
	return (0);
}

STATIC int
testcase_10(union ioctypes *ioc, mblk_t *dp)
{
	ioc->iocblk.ioc_rval = 0;
	ioc->iocblk.ioc_error = 0;
	ioc->iocblk.ioc_count = 0;
	return (0);
}

/* 
 *  -------------------------------------------------------------------------
 *
 *  PUT routines
 *
 *  -------------------------------------------------------------------------
 */

#ifdef LIS
#define streamscall _RP
#endif

STATIC streamscall int
nullmod_wput(queue_t *q, mblk_t *mp)
{
	switch (mp->b_datap->db_type) {
	case M_IOCTL:
	{
		int err = 0;
		union ioctypes *ioc;

		ioc = (typeof(ioc)) mp->b_rptr;
		if (_IOC_TYPE(ioc->iocblk.ioc_cmd) != 'V')
			break;
		switch (_IOC_NR(ioc->iocblk.ioc_cmd)) {
		case 1:
			err = testcase_1(ioc, mp->b_cont);
			break;
		case 2:
			err = testcase_2(ioc, mp->b_cont);
			break;
		case 3:
			err = testcase_3(ioc, mp->b_cont);
			break;
		case 4:
			err = testcase_4(ioc, mp->b_cont);
			break;
		case 5:
			err = testcase_5(ioc, mp->b_cont);
			break;
		case 6:
			err = testcase_6(ioc, mp->b_cont);
			break;
		case 7:
			err = testcase_7(ioc, mp->b_cont);
			break;
		case 8:
			err = testcase_8(ioc, mp->b_cont);
			break;
		case 9:
			err = testcase_9(ioc, mp->b_cont);
			break;
		case 10:
			err = testcase_10(ioc, mp->b_cont);
			break;
		default:
			err = -EINVAL;
			break;
		}
		if (err) {
			mp->b_datap->db_type = M_IOCNAK;
			ioc->iocblk.ioc_count = 0;
			ioc->iocblk.ioc_error = -err;
			ioc->iocblk.ioc_rval = -1;
		} else {
			mp->b_datap->db_type = M_IOCACK;
		}
		qreply(q, mp);
		return (0);
	}
	}
	putnext(q, mp);
	return (0);
}

STATIC streamscall int
nullmod_rput(queue_t *q, mblk_t *mp)
{
	putnext(q, mp);
	return (0);
}

/*
 *  -------------------------------------------------------------------------
 *
 *  OPEN and CLOSE
 *
 *  -------------------------------------------------------------------------
 */
STATIC streamscall int
nullmod_open(queue_t *q, dev_t *devp, int oflag, int sflag, cred_t *crp)
{
	queue_t *wq;

	if (q->q_ptr != NULL) {
		return (0);	/* already open */
	}
	wq = WR(q);
	if (sflag == MODOPEN || wq->q_next != NULL) {
		/* inherit packet sizes of what we are pushed over */
		wq->q_minpsz = wq->q_next->q_minpsz;
		wq->q_maxpsz = wq->q_next->q_maxpsz;
		q->q_ptr = wq->q_ptr = q;	/* just set it to something */
		qprocson(q);
		return (0);
	}
	return (EIO);		/* can't be opened as driver */
}
STATIC streamscall int
nullmod_close(queue_t *q, int oflag, cred_t *crp)
{
	(void) oflag;
	(void) crp;
	if (!q->q_ptr)
		return (ENXIO);
	qprocsoff(q);
	q->q_ptr = WR(q)->q_ptr = NULL;
	return (0);
}

/* 
 *  -------------------------------------------------------------------------
 *
 *  Registration and initialization
 *
 *  -------------------------------------------------------------------------
 */
STATIC struct qinit nullmod_rinit = {
	.qi_putp = nullmod_rput,
	.qi_qopen = nullmod_open,
	.qi_qclose = nullmod_close,
	.qi_minfo = &nullmod_minfo,
	.qi_mstat = &nullmod_rstat,
};

STATIC struct qinit nullmod_winit = {
	.qi_putp = nullmod_wput,
	.qi_minfo = &nullmod_minfo,
	.qi_mstat = &nullmod_wstat,
};

STATIC struct streamtab nullmod_info = {
	.st_rdinit = &nullmod_rinit,
	.st_wrinit = &nullmod_winit,
};

#ifdef LFS
STATIC struct fmodsw nullmod_fmod = {
	.f_name = CONFIG_STREAMS_NULLMOD_NAME,
	.f_str = &nullmod_info,
	.f_flag = D_MP,
	.f_kmod = THIS_MODULE,
};
#endif

#ifdef LFS
struct nullmod_ioctl {
	unsigned int cmd;
	void *opaque;
};

static struct nullmod_ioctl nullmod_map[] = {
	{.cmd = (('V' << 8) | 1),}
	, {.cmd = (('V' << 8) | 2),}
	, {.cmd = (('V' << 8) | 3),}
	, {.cmd = (('V' << 8) | 4),}
	, {.cmd = (('V' << 8) | 5),}
	, {.cmd = (('V' << 8) | 6),}
	, {.cmd = (('V' << 8) | 7),}
	, {.cmd = (('V' << 8) | 8),}
	, {.cmd = (('V' << 8) | 9),}
	, {.cmd = (('V' << 8) | 10),}
	, {.cmd = 0,}
};

static void
nullmod_unregister_ioctl32(void)
{
	struct nullmod_ioctl *i;

	for (i = nullmod_map; i->cmd != 0; i++)
		if (i->opaque != NULL)
			unregister_ioctl32(i->opaque);
}

static int
nullmod_register_ioctl32(void)
{
	struct nullmod_ioctl *i;

	for (i = nullmod_map; i->cmd != 0; i++) {
		if ((i->opaque = register_ioctl32(i->cmd)) == NULL) {
			nullmod_unregister_ioctl32();
			return (-ENOMEM);
		}
	}
	return (0);
}
#endif				/* LFS */

#ifdef CONFIG_STREAMS_NULLMOD_MODULE
STATIC
#endif
int __init
nullmod_init(void)
{
	int err;

#ifdef CONFIG_STREAMS_NULLMOD_MODULE
	printk(KERN_INFO NULLMOD_BANNER);
#else
	printk(KERN_INFO NULLMOD_SPLASH);
#endif
	nullmod_minfo.mi_idnum = modid;
#ifdef LFS
	if ((err = nullmod_register_ioctl32()) < 0)
		return (err);
	if ((err = register_strmod(&nullmod_fmod)) < 0) {
		nullmod_unregister_ioctl32();
		return (err);
	}
#endif
#ifdef LIS
	if ((err = lis_register_strmod(&nullmod_info, CONFIG_STREAMS_NULLMOD_NAME)) < 0)
		return (err);
#endif
	if (modid == 0 && err > 0)
		modid = err;
#ifdef LIS
	if ((err = lis_register_module_qlock_option(modid, LIS_QLOCK_NONE)) < 0) {
		lis_unregister_strmod(&nullmod_info);
		return (err);
	}
#endif
	return (0);
}

#ifdef CONFIG_STREAMS_NULLMOD_MODULE
STATIC
#endif
void __exit
nullmod_exit(void)
{
#ifdef LFS
	unregister_strmod(&nullmod_fmod);
	nullmod_unregister_ioctl32();
#endif
#ifdef LIS
	lis_unregister_strmod(&nullmod_info);
#endif
}

#ifdef CONFIG_STREAMS_NULLMOD_MODULE
module_init(nullmod_init);
module_exit(nullmod_exit);
#endif
